libass/ass_render.c

```diff
@@ -1850,40 +1854,46 @@ fix_glyph_scaling(ASS_Renderer *priv, GlyphInfo *glyph)
     glyph->font_size = ft_size;
 }
 
- /**
-  * \brief Checks whether a glyph should start a new bitmap run
-  * \param info Pointer to new GlyphInfo to check
-  * \param current_info Pointer to CombinedBitmapInfo for current run (may be NULL)
-  * \return true if a new run should be started
-  */
-static bool is_new_bm_run(GlyphInfo *info, GlyphInfo *last)
+// Initial run splitting based purely on the characters' styles
+static void split_style_runs(ASS_Renderer *render_priv)
 {
-    return !last || info->effect || info->drawing_text || last->drawing_text ||
-        strcmp(last->font->desc.family, info->font->desc.family) ||
-        last->font->desc.vertical != info->font->desc.vertical ||
-        last->font_size != info->font_size ||
-        last->c[0] != info->c[0] ||
-        last->c[1] != info->c[1] ||
-        last->c[2] != info->c[2] ||
-        last->c[3] != info->c[3] ||
-        last->be != info->be ||
-        last->blur != info->blur ||
-        last->shadow_x != info->shadow_x ||
-        last->shadow_y != info->shadow_y ||
-        last->frx != info->frx ||
-        last->fry != info->fry ||
-        last->frz != info->frz ||
-        last->fax != info->fax ||
-        last->fay != info->fay ||
-        last->scale_x != info->scale_x ||
-        last->scale_y != info->scale_y ||
-        last->border_style != info->border_style ||
-        last->border_x != info->border_x ||
-        last->border_y != info->border_y ||
-        last->hspacing != info->hspacing ||
-        last->italic != info->italic ||
-        last->bold != info->bold ||
-        ((last->flags ^ info->flags) & ~DECO_ROTATE);
+    for (int i = 0; i < render_priv->text_info.length; i++) {
+        GlyphInfo *info = render_priv->text_info.glyphs + i;
+        if (!i) {
+            info->starts_new_run = true;
```

Comment:
Is there a need for that? First run is handled by `bool new_run = true;` later.

Reply:
Not for bitmap combining, no. But it may make other future uses of this flag simpler: for one, it makes karaoke code a _tiny_ bit simpler. Also, for reference, karaoke will need an `if (!i)` section here in any case.

Also, the flag will anyway be set on the first glyph by `trim_whitespace` if the first line has no leading whitespace. So it wouldn’t be wise to have any code relying on the _absence_ of this flag for the first glyph, either.

But if you prefer to avoid setting this on the first glyph explicitly, I don’t mind removing this line and adding a `|| !i` in karaoke later. We’ll need to keep in mind that the flag _may or may not_ be set on the first glyph, then.

Reply:
I think it's better to keep `i = 0` case in this function but outside of this loop.

Reply:
I missed the fact that `text_info.length` is guaranteed to be nonzero. (This is checked just before the call to `split_style_runs`.)

I’ve now moved this assignment out of the loop.